ClusterRollingRestartTest.test_rollingRestart[clusterState:FROZEN,partitionAssignmentType:NEVER]

```
java.lang.AssertionError: Cluster size is not correct. Expected: 3 Actual: 1
	at org.junit.Assert.fail(Assert.java:88)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSize(HazelcastTestSupport.java:694)
	at com.hazelcast.test.HazelcastTestSupport$10.run(HazelcastTestSupport.java:713)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:835)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:709)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:704)
	at com.hazelcast.cluster.impl.ClusterRollingRestartTest.test_rollingRestart(ClusterRollingRestartTest.java:86)
```

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-IbmJDK1.8/78/com.hazelcast$hazelcast/testReport/com.hazelcast.cluster.impl/ClusterRollingRestartTest/test_rollingRestart_clusterState_FROZEN_partitionAssignmentType_NEVER_/

--- a/hazelcast/src/main/java/com/hazelcast/partition/impl/InternalPartitionServiceImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/partition/impl/InternalPartitionServiceImpl.java
@@ -86,11 +86,9 @@ import java.util.List;
 import java.util.Map;
 import java.util.Properties;
 import java.util.Set;
-import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.Future;
-import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.Semaphore;
 import java.util.concurrent.TimeUnit;
@@ -148,8 +146,8 @@ public class InternalPartitionServiceImpl implements InternalPartitionService, M
 
     @Probe
     private final AtomicInteger stateVersion = new AtomicInteger();
-    @Probe(name = "migrationQueueSize")
-    private final BlockingQueue<Runnable> migrationQueue = new LinkedBlockingQueue<Runnable>();
+
+    private final MigrationQueue migrationQueue = new MigrationQueue();
     private final AtomicBoolean migrationAllowed = new AtomicBoolean(true);
     @Probe(name = "lastRepartitionTime")
     private final AtomicLong lastRepartitionTime = new AtomicLong();
@@ -1254,8 +1252,8 @@ public class InternalPartitionServiceImpl implements InternalPartitionService, M
 
     @Override
     public boolean hasOnGoingMigrationLocal() {
-        return !activeMigrations.isEmpty() || !migrationQueue.isEmpty()
-                || migrationThread.isMigrating();
+        return !activeMigrations.isEmpty() || migrationQueue.isNonEmpty()
+                || migrationQueue.hasMigrationTasks();
     }
 
     private boolean isReplicaInSyncState() {
@@ -1650,6 +1648,7 @@ public class InternalPartitionServiceImpl implements InternalPartitionService, M
     }
 
     @Override
+    @Probe(name = "migrationQueueSize")
     public long getMigrationQueueSize() {
         return migrationQueue.size();
     }
@@ -1809,7 +1808,7 @@ public class InternalPartitionServiceImpl implements InternalPartitionService, M
         @Override
         public void run() {
             if (node.isMaster() && node.getState() == NodeState.ACTIVE) {
-                if (!migrationQueue.isEmpty() && isMigrationAllowed()) {
+                if (migrationQueue.isNonEmpty() && isMigrationAllowed()) {
                     logger.info("Remaining migration tasks in queue => " + migrationQueue.size());
                 }
                 publishPartitionRuntimeState();
@@ -1924,10 +1923,7 @@ public class InternalPartitionServiceImpl implements InternalPartitionService, M
         private void migratePartitionToNewOwner(int partitionId, Address[] replicas, Address currentOwner, Address newOwner) {
             MigrationInfo info = new MigrationInfo(partitionId, currentOwner, newOwner);
             MigrateTask migrateTask = new MigrateTask(info, replicas);
-            boolean offered = migrationQueue.offer(migrateTask);
-            if (!offered) {
-                logger.severe("Failed to offer: " + migrateTask);
-            }
+            migrationQueue.add(migrateTask);
         }
 
         private void assignNewPartitionOwner(int partitionId, Address[] replicas, InternalPartitionImpl currentPartition,
@@ -1947,7 +1943,7 @@ public class InternalPartitionServiceImpl implements InternalPartitionService, M
         }
     }
 
-    private class MigrateTask implements Runnable {
+    class MigrateTask implements Runnable {
         final MigrationInfo migrationInfo;
         final Address[] addresses;
 
@@ -2077,7 +2073,6 @@ public class InternalPartitionServiceImpl implements InternalPartitionService, M
 
     private class MigrationThread extends Thread implements Runnable {
         private final long sleepTime = max(250L, partitionMigrationInterval);
-        private volatile boolean migrating;
 
         MigrationThread(Node node) {
             super(node.getHazelcastThreadGroup().getInternalThreadGroup(),
@@ -2102,7 +2097,7 @@ public class InternalPartitionServiceImpl implements InternalPartitionService, M
         }
 
         private void doRun() throws InterruptedException {
-            for (; ; ) {
+            for (; ;) {
                 if (!isMigrationAllowed()) {
                     break;
                 }
@@ -2118,8 +2113,7 @@ public class InternalPartitionServiceImpl implements InternalPartitionService, M
             }
             boolean hasNoTasks = migrationQueue.isEmpty();
             if (hasNoTasks) {
-                if (migrating) {
-                    migrating = false;
+                if (!migrationQueue.hasMigrationTasks()) {
                     logger.info("All migration tasks have been completed, queues are empty.");
                 }
                 evictCompletedMigrations();
@@ -2130,15 +2124,18 @@ public class InternalPartitionServiceImpl implements InternalPartitionService, M
         }
 
         boolean processTask(Runnable r) {
-            if (r == null || isInterrupted()) {
-                return false;
-            }
-            migrating = (r instanceof MigrateTask);
             try {
+                if (r == null || isInterrupted()) {
+                    return false;
+                }
+
                 r.run();
             } catch (Throwable t) {
                 logger.warning(t);
+            } finally {
+                migrationQueue.afterTaskCompletion(r);
             }
+
             return true;
         }
 
@@ -2147,9 +2144,6 @@ public class InternalPartitionServiceImpl implements InternalPartitionService, M
             interrupt();
         }
 
-        boolean isMigrating() {
-            return migrating;
-        }
     }
 
     private static final class InternalPartitionListener implements PartitionListener {

--- a/hazelcast/src/main/java/com/hazelcast/partition/impl/MigrationQueue.java
+++ b/hazelcast/src/main/java/com/hazelcast/partition/impl/MigrationQueue.java
@@ -1,0 +1,82 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.partition.impl;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Manages migration tasks and migration status flag for {@link InternalPartitionServiceImpl} safely. Once a migration task
+ * is added to the queue, queue has to be notified via {@link MigrationQueue#afterTaskCompletion(Runnable)} after its execution.
+ */
+class MigrationQueue {
+
+    private final AtomicInteger migrateTaskCount = new AtomicInteger();
+
+    private final BlockingQueue<Runnable> queue = new LinkedBlockingQueue<Runnable>();
+
+    public void add(Runnable task) {
+        if (task instanceof InternalPartitionServiceImpl.MigrateTask) {
+            migrateTaskCount.incrementAndGet();
+        }
+
+        queue.offer(task);
+    }
+
+    public Runnable poll(int timeout, TimeUnit unit)
+            throws InterruptedException {
+        return queue.poll(timeout, unit);
+    }
+
+    public void clear() {
+        List<Runnable> sink = new ArrayList<Runnable>();
+        queue.drainTo(sink);
+
+        for (Runnable task : sink) {
+            afterTaskCompletion(task);
+        }
+    }
+
+    public void afterTaskCompletion(Runnable task) {
+        if (task instanceof InternalPartitionServiceImpl.MigrateTask) {
+            if (migrateTaskCount.decrementAndGet() < 0) {
+                throw new IllegalStateException();
+            }
+        }
+    }
+
+    public boolean hasMigrationTasks() {
+        return migrateTaskCount.get() > 0;
+    }
+
+    public boolean isEmpty() {
+        return queue.isEmpty();
+    }
+
+    public boolean isNonEmpty() {
+        return !queue.isEmpty();
+    }
+
+    public int size() {
+        return queue.size();
+    }
+
+}

--- a/hazelcast/src/test/java/com/hazelcast/partition/impl/MigrationQueueTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/partition/impl/MigrationQueueTest.java
@@ -1,0 +1,77 @@
+package com.hazelcast.partition.impl;
+
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+
+@Category(QuickTest.class)
+public class MigrationQueueTest {
+
+    private final MigrationQueue migrationQueue = new MigrationQueue();
+
+    @Test
+    public void test_migrationTaskCount_incremented_onMigrateTask() {
+        migrationQueue.add(mock(InternalPartitionServiceImpl.MigrateTask.class));
+
+        assertTrue(migrationQueue.hasMigrationTasks());
+        assertEquals(1, migrationQueue.size());
+    }
+
+    @Test
+    public void test_migrationTaskCount_notIncremented_onNonMigrateTask() {
+        migrationQueue.add(mock(Runnable.class));
+
+        assertFalse(migrationQueue.hasMigrationTasks());
+        assertEquals(1, migrationQueue.size());
+    }
+
+    @Test
+    public void test_migrationTaskCount_notDecremented_afterMigrateTaskPolled()
+            throws InterruptedException {
+        migrationQueue.add(mock(InternalPartitionServiceImpl.MigrateTask.class));
+        migrationQueue.poll(1, TimeUnit.SECONDS);
+
+        assertTrue(migrationQueue.hasMigrationTasks());
+    }
+
+    @Test
+    public void test_migrateTaskCount_decremented_afterMigrateTaskCompleted()
+            throws InterruptedException {
+        final InternalPartitionServiceImpl.MigrateTask migrateTask = mock(InternalPartitionServiceImpl.MigrateTask.class);
+
+        migrationQueue.add(migrateTask);
+        migrationQueue.afterTaskCompletion(migrateTask);
+
+        assertFalse(migrationQueue.hasMigrationTasks());
+    }
+
+    @Test
+    public void test_migrateTaskCount_notDecremented_afterNonMigrateTaskCompleted()
+            throws InterruptedException {
+        migrationQueue.add(mock(InternalPartitionServiceImpl.MigrateTask.class));
+        migrationQueue.afterTaskCompletion(mock(Runnable.class));
+
+        assertTrue(migrationQueue.hasMigrationTasks());
+    }
+
+    @Test
+    public void test_migrateTaskCount_decremented_onClear() {
+        migrationQueue.add(mock(InternalPartitionServiceImpl.MigrateTask.class));
+        migrationQueue.clear();
+
+        assertFalse(migrationQueue.hasMigrationTasks());
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void test_migrateTaskCount_notDecremented_belowZero() {
+        migrationQueue.afterTaskCompletion(mock(InternalPartitionServiceImpl.MigrateTask.class));
+    }
+
+}

